spark/v3.3/spark/src/main/scala/org/apache/spark/sql/execution/datasources/SparkExpressionConverter.scala

```diff
@@ -36,15 +35,30 @@ object SparkExpressionConverter {
     SparkFilters.convert(DataSourceStrategy.translateFilter(sparkExpression, supportNestedPredicatePushdown = true).get)
   }
 
-  @throws[AnalysisException]
-  def collectResolvedSparkExpression(session: SparkSession, tableName: String, where: String): Expression = {
+  def collectDeterministicSparkExpression(session: SparkSession,
+                                          tableName: String, where: String): Boolean = {
+    // used only to check if a deterministic expression is true or false
+    val tableAttrs = session.table(tableName).queryExecution.analyzed.output
+    val firstColumnName = tableAttrs.head.name
+    val anotherWhere = s"$firstColumnName is not null and $where"
+    val unresolvedExpression = session.sessionState.sqlParser.parseExpression(anotherWhere)
+    val filter = Filter(unresolvedExpression, DummyRelation(tableAttrs))
+    val optimizedLogicalPlan = session.sessionState.executePlan(filter).optimizedPlan
+    val option = optimizedLogicalPlan.collectFirst {
+      case filter: Filter => Some(filter.condition)
+    }.getOrElse(Option.empty)
+    if (option.isDefined) true else false
+  }
+
+  def collectResolvedSparkExpressionOption(session: SparkSession,
+                                           tableName: String, where: String): Option[Expression] = {
     val tableAttrs = session.table(tableName).queryExecution.analyzed.output
     val unresolvedExpression = session.sessionState.sqlParser.parseExpression(where)
     val filter = Filter(unresolvedExpression, DummyRelation(tableAttrs))
     val optimizedLogicalPlan = session.sessionState.executePlan(filter).optimizedPlan
     optimizedLogicalPlan.collectFirst {
-      case filter: Filter => filter.condition
-    }.getOrElse(throw new AnalysisException("Failed to find filter expression"))
+      case filter: Filter => Some(filter.condition)
```

Comment:
I assume we should have 3 branches:

```
case filter: Filter => filter.condition
case relation: DummyRelation => Literal.TrueLiteral
case relation: LocalRelation => Literal.FalseLiteral
```

Reply:
1st -> we have a real filter
2nd -> filter evaluates to true
3rd -> filter evaluates to false

Reply:
Hi @aokolnychyi , thanks for your suggestions.
I am trying to combine `collectResolvedSparkExpressionOption` and `collectDeterministicSparkExpression ` together. 
 could you give more explanation about your suggestions? 
thanks

Reply:
@ludlows, I think the underlying problem we are trying to solve is that the logic in the existing `collectFirst` is not exhaustive and does not cover cases when the condition evaluates to `true` or `false`. Instead of offering a new method to check if the condition evaluates to `true` or `false`, we should modify the existing method to take those scenarios into account.

I believe we only check this now:

```
case filter: Filter => filter.condition
```

I think it should be something like this instead:

```
case filter: Filter => filter.condition // the condition is non-trivial 
case relation: DummyRelation => Literal.TrueLiteral // the condition is always true and the filter was removed by the optimizer
case relation: LocalRelation => Literal.FalseLiteral // the condition is always false and the optimizer replaced the plan with an empty relation
```

Reply:
I think @aokolnychyi comment is the right way:

We just need to modify the method :  collectResolvedIcebergExpression  , and add those extra scala pattern-matching that Anton showed.  

Then, no need to modify the outside method.  The convert will automatically convert those Spark true/false to Iceberg true/false.

Reply:
Hi @ludlows , actually taking a look at this, why didn't we try just using @aokolnychyi 's code suggestion directly?  
```
  def collectResolvedSparkExpression(session: SparkSession, tableName: String, where: String): Expression = {
    val tableAttrs = session.table(tableName).queryExecution.analyzed.output
    val unresolvedExpression = session.sessionState.sqlParser.parseExpression(where)
    val filter = Filter(unresolvedExpression, DummyRelation(tableAttrs))
    val optimizedLogicalPlan = session.sessionState.executePlan(filter).optimizedPlan
    optimizedLogicalPlan.collectFirst {
      case filter: Filter => filter.condition // the condition is non-trivial
      case relation: DummyRelation => Literal.TrueLiteral // the condition is always true and the filter was removed by the optimizer 
      case relation: LocalRelation => Literal.FalseLiteral // the condition is always false and the optimizer replaced the plan with an empty relation
    }.getOrElse(throw new AnalysisException("Failed to find filter expression"))
  }
```
It looks like it should work.  ConstantFolding rule will probably get rid of the filter and make this match the second case (DummyRelation).  And we have added test case for alwaysTrue and alwaysFalse to catch if we miss something.

Reply:
hi @szehon-ho, yes. @aokolnychyi  and you are right.  now the current version is using this method to distinguish alwaysTrue, alwaysFalse and undetermined. thanks for your explanation. as I remember, I didn't use this method previously since  I didn't understand the behavior of `collectFirst` .